src/main/java/org/jabref/gui/customentrytypes/CustomEntryTypeDialogViewModel.java

```diff
@@ -0,0 +1,191 @@
+
+import org.fxmisc.easybind.EasyBind;
+
+public class CustomEntryTypeDialogViewModel {
+
+    public static final StringConverter<Field> FIELD_STRING_CONVERTER = new StringConverter<>() {
+
+        @Override
+        public String toString(Field object) {
+            return object != null ? object.getDisplayName() : "";
+        }
+
+        @Override
+        public Field fromString(String string) {
+            return new UnknownField(string);
+        }
+    };
+
+    private ListProperty<BibEntryType> entryTypesProperty;
+    private ListProperty<Field> fieldsProperty;
+    private ObjectProperty<BibEntryType> selectedEntryTypesProperty = new SimpleObjectProperty<>();
+    private ListProperty<FieldViewModel> fieldsForTypeProperty;
+    private ObjectProperty<Field> selectedFieldToAddProperty = new SimpleObjectProperty<>();
+    private StringProperty entryTypeToAddProperty = new SimpleStringProperty("");
+    private ObservableList<BibEntryType> entryTypes;
+    private ObservableList<FieldViewModel> fieldsForType = FXCollections.observableArrayList(extractor -> new Observable[] {extractor.fieldNameProperty(), extractor.fieldTypeProperty()});
+    private ObjectProperty<Field> newFieldToAddProperty = new SimpleObjectProperty<>();
+    private BibDatabaseMode mode;
+    private Map<BibEntryType, List<FieldViewModel>> typesWithFields = new HashMap<>();
+    private List<BibEntryType> typesToRemove = new ArrayList<>();
+
+    private PreferencesService preferencesService;
+
+    public CustomEntryTypeDialogViewModel(BibDatabaseMode mode, PreferencesService preferencesService) {
+        this.mode = mode;
+        this.preferencesService = preferencesService;
+
+        Globals.entryTypesManager.addCustomOrModifiedTypes(preferencesService.loadBibEntryTypes(BibDatabaseMode.BIBTEX),
+                                                           preferencesService.loadBibEntryTypes(BibDatabaseMode.BIBLATEX));
+
+        Collection<BibEntryType> allTypes = Globals.entryTypesManager.getAllTypes(mode);
+        allTypes.addAll(Globals.entryTypesManager.getAllCustomTypes(mode));
+
+        entryTypes = FXCollections.observableArrayList(allTypes);
+        entryTypesProperty = new SimpleListProperty<>(entryTypes);
+
+        fieldsProperty = new SimpleListProperty<>(FXCollections.observableArrayList(FieldFactory.getAllFields()));
+
+        for (BibEntryType entryType : allTypes) {
+            List<FieldViewModel> fields = entryType.getAllFields().stream().map(bibField -> new FieldViewModel(bibField.getField(), entryType.isRequired(bibField.getField()), bibField.getPriority(), entryType)).collect(Collectors.toList());
+            typesWithFields.put(entryType, fields);
+        }
+
+        this.fieldsForTypeProperty = new SimpleListProperty<>(fieldsForType);
+
+        EasyBind.subscribe(selectedEntryTypesProperty, type -> {
+            if (type != null) {
+                List<FieldViewModel> typesForField = typesWithFields.get(type);
```

Comment:
`typesForField` seems to be a misnomer as its the other way around. Variable can be inlined anyway in my opinion